File: apps/mobile/lib/core/logic/recent_movers.dart
import '../models/binder_entry.dart';

/// Observed start Low vs latest catalog Low for one Printing, plus the derived
/// percent and amount. Ranking uses [percentChange], never the currency amount.
class RecentLowChange {
  const RecentLowChange({
    required this.cardId,
    required this.startOn,
    required this.startLow,
    required this.latestLow,
    required this.percentChange,
    required this.amountChange,
  });

  final String cardId;

  /// Catalog calendar date (`YYYY-MM-DD`) of the start snapshot.
  final String startOn;

  /// Observed marketplace Low on [startOn]. Never invented.
  final double startLow;

  /// Observed current catalog Low. Never invented.
  final double latestLow;

  /// `(latestLow - startLow) / startLow`.
  final double percentChange;

  /// `latestLow - startLow`.
  final double amountChange;

  Map<String, dynamic> toJson() => {
        'card_id': cardId,
        'start_on': startOn,
        'start_low': startLow,
        'latest_low': latestLow,
        'percent_change': percentChange,
        'amount_change': amountChange,
      };

  factory RecentLowChange.fromMap(Map<String, dynamic> map) {
    double requireLow(String key) {
      final value = map[key];
      if (value is num) return value.toDouble();
      throw FormatException('Recent change $key must be a number, got $value');
    }

    return RecentLowChange(
      cardId: map['card_id'] as String,
      startOn: (map['start_on'] as Object).toString(),
      startLow: requireLow('start_low'),
      latestLow: requireLow('latest_low'),
      percentChange: requireLow('percent_change'),
      amountChange: requireLow('amount_change'),
    );
  }
}

/// One row from `fab_recent_movers`. Identity fields come from the catalog
/// join; Low fields are observed snapshots, not interpolated.
class RecentMoverRow {
  const RecentMoverRow({
    required this.direction,
    required this.rank,
    required this.cardId,
    required this.name,
    required this.setName,
    required this.finish,
    this.imageUrl,
    required this.startLow,
    required this.startOn,
    required this.latestLow,
    required this.percentChange,
    required this.amountChange,
  });

  final String direction;
  final int rank;
  final String cardId;
  final String name;
  final String setName;
  final String finish;
  final String? imageUrl;
  final double startLow;
  final String startOn;
  final double latestLow;
  final double percentChange;
  final double amountChange;

  bool get isGainer => direction == 'gainer';

  factory RecentMoverRow.fromMap(Map<String, dynamic> map) {
    double requireLow(String key) {
      final value = map[key];
      if (value is num) return value.toDouble();
      throw FormatException('Recent mover $key must be a number, got $value');
    }

    final direction = map['direction'] as String? ?? '';
    if (direction != 'gainer' && direction != 'loser') {
      throw FormatException('Recent mover direction must be gainer or loser');
    }

    return RecentMoverRow(
      direction: direction,
      rank: (map['rank'] as num).toInt(),
      cardId: map['card_id'] as String,
      name: map['name'] as String? ?? '',
      setName: map['set_name'] as String? ?? '',
      finish: map['finish'] as String? ?? '',
      imageUrl: map['image_url'] as String?,
      startLow: requireLow('start_low'),
      startOn: (map['start_on'] as Object).toString(),
      latestLow: requireLow('latest_low'),
      percentChange: requireLow('percent_change'),
      amountChange: requireLow('amount_change'),
    );
  }
}

void _assertSource(String source) {
  switch (source) {
    case 'tcgplayer':
    case 'cardmarket':
      return;
    default:
      throw ArgumentError.value(
        source,
        'source',
        'must be tcgplayer or cardmarket',
      );
  }
}

double? _lowForSource(String source, Map<String, dynamic> row) {
  _assertSource(source);
  switch (source) {
    case 'tcgplayer':
      return _asDouble(row['tcg_low']);
    case 'cardmarket':
      return _asDouble(row['cm_low']);
    default:
      throw ArgumentError.value(
        source,
        'source',
        'must be tcgplayer or cardmarket',
      );
  }
}

double? _asDouble(Object? value) {
  if (value == null) return null;
  if (value is num) return value.toDouble();
  return null;
}

const double _startLowFloor = 1;
const double _maxObservedLow = 10000;
const double _maxAbsPercentChange = 10;

DateTime _parseDate(String ymd) {
  final parts = ymd.split('-');
  return DateTime(
    int.parse(parts[0]),
    int.parse(parts[1]),
    int.parse(parts[2]),
  );
}

String _formatDate(DateTime d) =>
    '${d.year.toString().padLeft(4, '0')}-'
    '${d.month.toString().padLeft(2, '0')}-'
    '${d.day.toString().padLeft(2, '0')}';

/// Inclusive start-Low window: `[today - 5 days, today - 3 days]`.
({String from, String to}) startLowWindow(String today) {
  final date = _parseDate(today);
  return (
    from: _formatDate(date.subtract(const Duration(days: 5))),
    to: _formatDate(date.subtract(const Duration(days: 3))),
  );
}

/// Observed start Low vs latest Low for one Printing. No floor, no sealed
/// filter, no rank cut — those belong to [rankRecentMovers] only.
RecentLowChange? observedPrintingRecentChange({
  required String today,
  required String source,
  required List<Map<String, dynamic>> snapshots,
  required Map<String, dynamic>? current,
  required String cardId,
}) {
  if (current == null) return null;
  final latestLow = _lowForSource(source, current);
  if (latestLow == null) return null;

  final window = startLowWindow(today);
  String? startOn;
  double? startLow;
  for (final snap in snapshots) {
    final capturedOn = snap['captured_on'] as String?;
    if (capturedOn == null) continue;
    if (capturedOn.compareTo(window.from) < 0) continue;
    if (capturedOn.compareTo(window.to) > 0) continue;
    final low = _lowForSource(source, snap);
    if (low == null) continue;
    if (startOn == null || capturedOn.compareTo(startOn) > 0) {
      startOn = capturedOn;
      startLow = low;
    }
  }
  if (startOn == null || startLow == null) return null;
  if (startLow > _maxObservedLow || latestLow > _maxObservedLow) return null;
  if (latestLow == startLow) return null;

  final amountChange = latestLow - startLow;
  final percentChange = amountChange / startLow;
  if (percentChange.abs() > _maxAbsPercentChange) return null;
  return RecentLowChange(
    cardId: cardId,
    startOn: startOn,
    startLow: startLow,
    latestLow: latestLow,
    percentChange: percentChange,
    amountChange: amountChange,
  );
}

/// Rank eligible Printings by percent change of observed Low.
///
/// [snapshotsByCard] maps Printing id → history rows (`captured_on`, `tcg_low`,
/// `cm_low`). [currentLows] maps Printing id → `{tcg_low, cm_low, is_sealed}`.
/// Null Lows are omitted, never treated as zero.
({List<RecentLowChange> gainers, List<RecentLowChange> losers}) rankRecentMovers({
  required String today,
  required String source,
  required Map<String, List<Map<String, dynamic>>> snapshotsByCard,
  required Map<String, Map<String, dynamic>> currentLows,
}) {
  _assertSource(source);
  final eligible = <RecentLowChange>[];

  currentLows.forEach((cardId, current) {
    if (current['is_sealed'] == true) return;
    final change = observedPrintingRecentChange(
      today: today,
      source: source,
      snapshots: snapshotsByCard[cardId] ?? const <Map<String, dynamic>>[],
      current: current,
      cardId: cardId,
    );
    if (change == null) return;
    if (change.startLow < _startLowFloor) return;
    eligible.add(change);
  });

  int byCardId(RecentLowChange a, RecentLowChange b) =>
      a.cardId.compareTo(b.cardId);

  final gainers = eligible.where((r) => r.percentChange > 0).toList()
    ..sort((a, b) {
      final byPercent = b.percentChange.compareTo(a.percentChange);
      return byPercent != 0 ? byPercent : byCardId(a, b);
    });
  final losers = eligible.where((r) => r.percentChange < 0).toList()
    ..sort((a, b) {
      final byPercent = a.percentChange.compareTo(b.percentChange);
      return byPercent != 0 ? byPercent : byCardId(a, b);
    });

  return (
    gainers: gainers.take(10).toList(),
    losers: losers.take(10).toList(),
  );
}

/// Displayable recent Low change for requested Printing ids. No floor, no
/// top-10, no sealed extra-filter. Ids not in [cardIds] are ignored.
List<RecentLowChange> lookupPrintingRecentChanges({
  required String today,
  required String source,
  required Map<String, List<Map<String, dynamic>>> snapshotsByCard,
  required Map<String, Map<String, dynamic>> currentLows,
  required List<String> cardIds,
}) {
  _assertSource(source);
  final seen = <String>{};
  final overlays = <RecentLowChange>[];
  for (final cardId in cardIds) {
    if (!seen.add(cardId)) continue;
    final change = observedPrintingRecentChange(
      today: today,
      source: source,
      snapshots: snapshotsByCard[cardId] ?? const <Map<String, dynamic>>[],
      current: currentLows[cardId],
      cardId: cardId,
    );
    if (change != null) overlays.add(change);
  }
  return overlays;
}

String? _printingId(Object entry) {
  if (entry is BinderEntry) return entry.card.id;
  if (entry is Map) {
    final cardId = entry['cardId'] ?? entry['card_id'];
    if (cardId is String && cardId.isNotEmpty) return cardId;
    final card = entry['card'];
    if (card is Map && card['id'] is String) return card['id'] as String;
  }
  return null;
}

bool _isWanted(Object entry) {
  if (entry is BinderEntry) return entry.isWanted;
  if (entry is Map) return entry['isWanted'] == true || entry['is_wanted'] == true;
  return false;
}

int _quantity(Object entry) {
  if (entry is BinderEntry) return entry.quantity;
  if (entry is Map) {
    final qty = entry['quantity'];
    if (qty is num) return qty.toInt();
  }
  return 0;
}

bool _isTombstoned(Object entry) {
  if (entry is Map) {
    final deleted = entry['deletedAt'] ?? entry['deleted_at'];
    return deleted != null && deleted.toString().isNotEmpty;
  }
  return false;
}

/// Sums quantity per Printing id for a **pre-filtered** Binder list.
/// Display only — not a rank key and not sent to SQL.
///
/// Callers MUST filter to the open Binder first. Do not change
/// [ownedPrintingIds] to take a binder id.
Map<String, int> copiesByPrintingId(Iterable<Object> entries) {
  final copies = <String, int>{};
  for (final entry in entries) {
    if (_isWanted(entry) || _isTombstoned(entry)) continue;
    final qty = _quantity(entry);
    if (qty <= 0) continue;
    final id = _printingId(entry);
    if (id == null) continue;
    copies[id] = (copies[id] ?? 0) + qty;
  }
  return copies;
}

/// Distinct owned Printing ids: qty > 0 in any Binder, not Want List, not
/// tombstoned. Same Printing in two Binders or two conditions is one id.
List<String> ownedPrintingIds(Iterable<Object> entries) {
  final ids = <String>{};
  for (final entry in entries) {
    if (_isWanted(entry) || _isTombstoned(entry) || _quantity(entry) <= 0) {
      continue;
    }
    final id = _printingId(entry);
    if (id != null) ids.add(id);
  }
  return ids.toList();
}

File: apps/web/src/utils/recentMovers.js
/**
 * Recent movers ranking helpers. Production lists come from SQL
 * `fab_recent_movers`; this module exists so `packages/contracts/recent_movers.json`
 * can fail a build when JS drifts from Dart / SQL.
 *
 * Names distinguish observed start Low, observed latest Low, and derived
 * percent/amount. Null Lows are omitted, never treated as 0.
 */

const START_LOW_FLOOR = 1;
const MAX_OBSERVED_LOW = 10000;
const MAX_ABS_PERCENT_CHANGE = 10;
const MAX_MOVERS_PER_DIRECTION = 10;

const lowColumnForSource = {
    tcgplayer: 'tcg_low',
    cardmarket: 'cm_low',
};

function assertSource(source) {
    if (source !== 'tcgplayer' && source !== 'cardmarket') {
        throw new Error(`recent movers source must be tcgplayer or cardmarket, got ${source}`);
    }
}

function asNumber(value) {
    if (value == null || value === '') return null;
    const n = typeof value === 'number' ? value : Number(value);
    return Number.isFinite(n) ? n : null;
}

function lowForSource(source, row) {
    assertSource(source);
    return asNumber(row?.[lowColumnForSource[source]]);
}

function shiftDate(ymd, days) {
    const [year, month, day] = ymd.split('-').map(Number);
    const date = new Date(Date.UTC(year, month - 1, day));
    date.setUTCDate(date.getUTCDate() + days);
    return date.toISOString().slice(0, 10);
}

/** Inclusive start-Low window: `[today - 5 days, today - 3 days]`. */
export function startLowWindow(today) {
    return {
        from: shiftDate(today, -5),
        to: shiftDate(today, -3),
    };
}

function compareCardId(a, b) {
    if (a.card_id < b.card_id) return -1;
    if (a.card_id > b.card_id) return 1;
    return 0;
}

/**
 * Observed start Low vs latest Low for one Printing. No floor, no sealed
 * filter, no rank cut — those belong to [rankRecentMovers] only.
 */
function observedPrintingRecentChange({ today, source, snapshots, current, cardId }) {
    if (current == null) return null;
    const latestLow = lowForSource(source, current);
    if (latestLow == null) return null;

    const { from, to } = startLowWindow(today);
    let startOn = null;
    let startLow = null;
    for (const snap of snapshots || []) {
        const capturedOn = snap?.captured_on;
        if (!capturedOn || capturedOn < from || capturedOn > to) continue;
        const low = lowForSource(source, snap);
        if (low == null) continue;
        if (startOn == null || capturedOn > startOn) {
            startOn = capturedOn;
            startLow = low;
        }
    }
    if (startOn == null || startLow == null) return null;
    if (startLow > MAX_OBSERVED_LOW || latestLow > MAX_OBSERVED_LOW) return null;
    if (latestLow === startLow) return null;

    const amountChange = latestLow - startLow;
    const percentChange = amountChange / startLow;
    if (Math.abs(percentChange) > MAX_ABS_PERCENT_CHANGE) return null;
    return {
        card_id: cardId,
        start_on: startOn,
        start_low: startLow,
        latest_low: latestLow,
        percent_change: percentChange,
        amount_change: amountChange,
    };
}

/**
 * Rank eligible Printings by percent change of observed Low.
 *
 * @param {object} params
 * @param {string} params.today YYYY-MM-DD
 * @param {'tcgplayer'|'cardmarket'} params.source
 * @param {Record<string, Array<object>>} params.snapshotsByCard
 * @param {Record<string, object>} params.currentLows `{ tcg_low, cm_low, is_sealed }`
 */
export function rankRecentMovers({ today, source, snapshotsByCard, currentLows }) {
    assertSource(source);
    const eligible = [];

    for (const [cardId, current] of Object.entries(currentLows || {})) {
        if (current?.is_sealed) continue;
        const change = observedPrintingRecentChange({
            today,
            source,
            snapshots: snapshotsByCard?.[cardId],
            current,
            cardId,
        });
        if (!change) continue;
        if (change.start_low < START_LOW_FLOOR) continue;
        eligible.push(change);
    }

    const gainers = eligible
        .filter((row) => row.percent_change > 0)
        .sort((a, b) => (b.percent_change - a.percent_change) || compareCardId(a, b))
        .slice(0, MAX_MOVERS_PER_DIRECTION);
    const losers = eligible
        .filter((row) => row.percent_change < 0)
        .sort((a, b) => (a.percent_change - b.percent_change) || compareCardId(a, b))
        .slice(0, MAX_MOVERS_PER_DIRECTION);

    return { gainers, losers };
}

/**
 * Displayable recent Low change for requested Printing ids. No floor, no
 * top-10, no sealed extra-filter. Ids not in [cardIds] are ignored.
 *
 * @param {object} params
 * @param {string} params.today YYYY-MM-DD
 * @param {'tcgplayer'|'cardmarket'} params.source
 * @param {Record<string, Array<object>>} params.snapshotsByCard
 * @param {Record<string, object>} params.currentLows
 * @param {string[]} params.cardIds
 */
export function lookupPrintingRecentChanges({
    today,
    source,
    snapshotsByCard,
    currentLows,
    cardIds,
}) {
    assertSource(source);
    if (cardIds == null) {
        throw new Error('lookupPrintingRecentChanges: cardIds must not be null');
    }
    const seen = new Set();
    const overlays = [];
    for (const cardId of cardIds) {
        if (seen.has(cardId)) continue;
        seen.add(cardId);
        const change = observedPrintingRecentChange({
            today,
            source,
            snapshots: snapshotsByCard?.[cardId],
            current: currentLows?.[cardId],
            cardId,
        });
        if (change) overlays.push(change);
    }
    return overlays;
}

function printingId(entry) {
    return entry?.cardId || entry?.card_id || entry?.card?.id || null;
}

/**
 * Sums quantity per Printing id for a **pre-filtered** Binder list.
 * Display only — not a rank key and not sent to SQL.
 *
 * Callers MUST filter to the open Binder first. Do not change
 * [ownedPrintingIds] to take a binder id.
 */
export function copiesByPrintingId(entries) {
    const copies = new Map();
    for (const entry of entries || []) {
        if (entry?.isWanted || entry?.is_wanted) continue;
        if (entry?.deletedAt || entry?.deleted_at) continue;
        const qty = Number(entry?.quantity) || 0;
        if (qty <= 0) continue;
        const id = printingId(entry);
        if (!id) continue;
        copies.set(id, (copies.get(id) || 0) + qty);
    }
    return Object.fromEntries(copies);
}

/**
 * Distinct owned Printing ids: qty > 0 in any Binder, not Want List, not
 * tombstoned. Same Printing in two Binders or two conditions is one id.
 */
export function ownedPrintingIds(entries) {
    const ids = new Set();
    for (const entry of entries || []) {
        if (entry?.isWanted || entry?.is_wanted) continue;
        if (entry?.deletedAt || entry?.deleted_at) continue;
        const qty = Number(entry?.quantity) || 0;
        if (qty <= 0) continue;
        const id = printingId(entry);
        if (id) ids.add(id);
    }
    return [...ids];
}

export function splitMoversByDirection(rows) {
    const gainers = [];
    const losers = [];
    for (const row of rows || []) {
        if (row?.direction === 'gainer') gainers.push(row);
        else if (row?.direction === 'loser') losers.push(row);
    }
    return { gainers, losers };
}
